# Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio  test. The function should return the F-value and the corresponding p-value for the test.


In [3]:
from scipy.stats import f_oneway

def calculate_f_value(data1, data2):
    # Perform variance ratio test
    f_value, p_value = f_oneway(data1, data2)

    return f_value, p_value


# Example usage
data1 = [1, 2, 3, 4, 5]
data2 = [2, 4, 6, 8, 10]

f_value, p_value = calculate_f_value(data1, data2)
print("F-value:", f_value)
print("p-value:", p_value)


F-value: 3.6
p-value: 0.0943497728424377


# Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an  F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [4]:
from scipy.stats import f

def get_critical_f_value(significance_level, df_num, df_den):
    alpha = significance_level / 2  # Divide alpha by 2 for a two-tailed test
    critical_value = f.ppf(1 - alpha, df_num, df_den)
    return critical_value


significance_level = 0.05
df_num = 3
df_den = 20

critical_f_value = get_critical_f_value(significance_level, df_num, df_den)
print(critical_f_value)


3.8586986662732143


# Q3. Write a Python program that generates random samples from two normal distributions with known  variances and uses an F-test to determine if the variances are equal. The program should output the F value, degrees of freedom, and p-value for the test.

In [5]:
import numpy as np
from scipy.stats import f

def f_test(sample1, sample2, var1, var2):
    n1 = len(sample1)
    n2 = len(sample2)

    # Calculate the F-statistic
    f_value = np.var(sample1) / np.var(sample2)

    # Calculate the degrees of freedom
    df1 = n1 - 1
    df2 = n2 - 1

    # Calculate the p-value
    p_value = 2 * min(f.cdf(f_value, df1, df2), 1 - f.cdf(f_value, df1, df2))

    return f_value, df1, df2, p_value

# Generate random samples from two normal distributions
np.random.seed(42)  # Set a seed for reproducibility
sample1 = np.random.normal(loc=0, scale=1, size=100)
sample2 = np.random.normal(loc=0, scale=1.5, size=100)

# Known variances
var1 = 1
var2 = 1.5

# Perform the F-test
f_value, df1, df2, p_value = f_test(sample1, sample2, var1, var2)

# Print the results
print("F-value:", f_value)
print("Degrees of freedom:", df1, ",", df2)
print("P-value:", p_value)


F-value: 0.40304633927638567
Degrees of freedom: 99 , 99
P-value: 9.191624675736606e-06


# Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from  each population. Conduct an F-test at the 5% significance level to determine if the variances are  significantly different.




Null Hypothesis (H0): The variances are equal.
Alternative Hypothesis (Ha): The variances are significantly different.

 the significance level is given as 5%, which means α = 0.05.

F = s1² / s2²

where s1² is the larger sample variance and s2² is the smaller sample variance.

In this case, s1² = 15 (larger variance) and s2² = 10 (smaller variance).

F = 15 / 10 = 1.5

df1 = n1 - 1 = 12 - 1 = 11
df2 = n2 - 1 = 12 - 1 = 11

Using the significance level α = 0.05 and the degrees of freedom, we can find the critical value from an F-distribution table or using statistical software. For the given degrees of freedom, the critical value is approximately 3.49.

Step 5: Compare the calculated F-statistic with the critical value:
If the calculated F-statistic is greater than the critical value, we reject the null hypothesis and conclude that the variances are significantly different. Otherwise, we fail to reject the null hypothesis.

# Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25  products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance  level to determine if the claim is justified.


The null hypothesis (H₀) assumes that the claim is justified, stating that the population variance (σ²) is equal to the claimed variance (0.005):
H₀: σ² = 0.005

The aσ²) is greater than the claimed variance (0.005):
H₁: σ² > 0.005


F = Sample Variance 1 / Sample Variance 2




Calculating the degrees of freedom:
df₁ = n₁ - 1 = 25 - 1 = 24
df₂ = n₂ - 1 = 25 - 1 = 24

Now, we can calculate the F-value:
F = Sample Variance 1 / Sample Variance 2
F = 0.006 / 0.005
F = 1.2

Assuming the critical F-value at the 1% significance level is F and if the calculated F-value is greater than F_c, we reject the null hypothesis


# Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an  F-distribution and calculates the mean and variance of the distribution. The function should return the  mean and variance as a tuple.

In [7]:
def calculate_f_distribution_mean_variance(df_numerator, df_denominator):
    if df_numerator <= 0 or df_denominator <= 0:
        raise ValueError("Degrees of freedom must be greater than zero.")

    mean = df_denominator / (df_denominator - 2)

    if df_denominator <= 4:
        variance = float('inf')
    else:
        variance = (2 * (df_denominator ** 2) * (df_numerator + df_denominator - 2)) / (df_numerator * (df_denominator - 2) ** 2 * (df_denominator - 4))

    return mean, variance


df_numerator = 5
df_denominator = 10

mean, variance = calculate_f_distribution_mean_variance(df_numerator, df_denominator)
print("Mean:", mean)
print("Variance:", variance)


Mean: 1.25
Variance: 1.3541666666666667


# Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The  sample variance is found to be 25. Another random sample of 15 measurements is taken from another  normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test  at the 10% significance level to determine if the variances are significantly different.



Sample 1: n1 = 10 measurements, s1^2 = 25
Sample 2: n2 = 15 measurements, s2^2 = 20


F = s1^2 / s2^2

Under the null hypothesis of equal variances, the test statistic follows an F-distribution with (n1 - 1) and (n2 - 1) degrees of freedom.
.

If F > F_critical, we reject the null hypothesis and conclude that the variances are significantly different.


F = s1^2 / s2^2 = 25 / 20 = 1.25

If F > F_critical, we reject the null hypothesis. Otherwise, we fail to reject the null hypothesis.


# Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday  night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5%  significance level to determine if the variances are significantly different.

H0: The variances of the two samples are equal.
H1: The variances of the two samples are significantly different.


F = (s1^2) / (s2^2)


Step 5: Compare the test statistic (F) with the critical value. If the test statistic is greater than the critical value, we reject the null hypothesis; otherwise, we fail to reject the null hypothesis.


Sample size (n1) = 7
Mean (x̄1) = (24 + 25 + 28 + 23 + 22 + 20 + 27) / 7 = 23.86
Variance (s1^2) = [(24 - 23.86)^2 + (25 - 23.86)^2 + (28 - 23.86)^2 + (23 - 23.86)^2 + (22 - 23.86)^2 + (20 - 23.86)^2 + (27 - 23.86)^2] / (7 - 1) = 7.90


Sample size (n2) = 6
Mean (x̄2) = (31 + 33 + 35 + 30 + 32 + 36) / 6 = 32.83
Variance (s2^2) = [(31 - 32.83)^2 + (33 - 32.83)^2 + (35 - 32.83)^2 + (30 - 32.83)^2 + (32 - 32.83)^2 + (36 - 32.83)^2] / (6 - 1) = 6.47

Calculating the test statistic:
F = (s1^2) / (s2^2) = 7.90 / 6.47 = 1.22 (rounded to two decimal places)

 at α = 0.05, with df1 = 6 and df2 = 5, the critical value is approximately 5.99.

Step 5: Comparing the test statistic and critical value:
Since the test statistic (1.22) is less than the critical value (5.99), we fail to reject the null hypothesis.



# Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83;  Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances  are significantly different.



Null hypothesis (H₀): The variances of Group A and Group B are equal.
Alternative hypothesis (H₁): The variances of Group A and Group B are not equal.


Mean (x̄₁) = (80 + 85 + 90 + 92 + 87 + 83) / 6 = 87.83
Variance (s₁²) = Σ(xi - x̄₁)² / (n₁ - 1)
= ((80 - 87.83)² + (85 - 87.83)² + (90 - 87.83)² + (92 - 87.83)² + (87 - 87.83)² + (83 - 87.83)²) / (6 - 1)
= (46.39 + 8.03 + 5.34 + 16.54 + 0.70 + 21.47) / 5
= 18.47


Mean (x̄₂) = (75 + 78 + 82 + 79 + 81 + 84) / 6 = 80.83
Variance (s₂²) = Σ(xi - x̄₂)² / (n₂ - 1)
= ((75 - 80.83)² + (78 - 80.83)² + (82 - 80.83)² + (79 - 80.83)² + (81 - 80.83)² + (84 - 80.83)²) / (6 - 1)
= (33.77 + 8.11 + 1.39 + 2.47 + 0.03 + 9.11) / 5
= 10.18


 1% significance level to be approximately 8.70.


Compare the test statistic with the critical value.
Since the test statistic (F = 1.812) is less than the critical value (8.70), we do not reject the null hypothesis.
